README.md
# Latent Factor Models Based on Linear Total Correlation Explanation (CorEx)

Linear CorEx finds latent factors that are as informative as possible about relationships in the data. 
The approach is described in this paper:
[Low Complexity Gaussian Latent Factor Models and a Blessing of Dimensionality](https://arxiv.org/abs/1706.03353).
This is useful for covariance estimation, clustering related variables, and dimensionality reduction, especially 
in the high-dimensional, under-sampled regime. 

To install:
```
pip install linearcorex
```

Mathematically, the objective is to find factors, y, where y = W x and  
x in R^n is the data and W is an m by n weight matrix. 
We are minimizing TC(X|Y) + TC(Y) where TC is the "total correlation" or multivariate mutual information. This objective
is optimized when X's are independent after conditioning on Y's, and the Y's themselves are independent. 
Instead of heuristically upper bounding this objective as we do for discrete CorEx, 
we are able to optimize it exactly in the linear case. 
While this extension required assumptions of linearity, the 
advantage is that the code is pretty fast since it only relies on matrix algebra. In principle it could be 
further accelerated using GPUs. 


Without further constraints, the optima of this objective 
may have an undesirable property: information about the X_i's can be stored "synergistically" in the latent factors. 
In other words, to predict a single variable you need to combine info from all the latent factors. Therefore, we 
add a constraint that the solutions should be non-synergistic (latent factors are individually informative about each variable X_i). 
This also recovers the property of the original lower bound formulation from AISTATS that each latent factor
has a non-negative added contribution towards TC.
Note that by default, we constrain solutions to eliminate synergy. 
But, you can turn it off by setting eliminate_synergy=False in the python API or -a from the command line. 
For making nice trees, it should be left on (e.g. for personality data or ADNI data). 

To test the command line interface, try:
```
cd $INSTALL_DIRECTORY/linearcorex/
python vis_corex.py ../tests/data/test_big5.csv --layers=5,1 --verbose=1 --no_row_names -o big5
python vis_corex.py ../tests/data/adni_blood.csv --layers=30,5,1 --missing=-1e6 --verbose=1 -o adni
python vis_corex.py ../tests/data/matrix.tcga_ov.geneset1.log2.varnorm.RPKM.txt --layers=30,5,1 --delimiter=' ' --verbose=1 --gaussianize="outliers" -o gene
```
Each of these examples generates pairwise plots of relationships and a graph. 

The python API uses the sklearn conventions of fit/transform.  
```python
import linearcorex as lc
import numpy as np

out = lc.Corex(n_hidden=5, verbose=True)  # A Corex model with 5 factors
X = np.random.random((100, 50))  # Random data with 100 samples and 50 variables
out.fit(X)  # Fit the model on data
y = out.transform(X)  # Transform data into latent factors
print(out.clusters)  # See the clusters
cov = out.get_covariance()  # The covariance matrix
```


Missing values can be specified, but are just imputed in a naive way. 

## Papers

See [Sifting Common Info...](https://arxiv.org/abs/1606.02307) and 
[Maximally informative representations...](https://arxiv.org/abs/1410.7404) for work building up to this method. 
The main paper describing the method is 
[Low Complexity Gaussian Latent Factor Models and a Blessing of Dimensionality](https://arxiv.org/abs/1706.03353).
The connections with the idea of "synergy" will be described in future work. 


### Troubleshooting visualization
For Mac users: 

To get the visualization of the hierarchy looking nice sometimes takes a little effort. To get graphs to compile correctly do the following. 
Using "brew" to install, you need to do "brew install gts" followed by "brew install --with-gts graphviz". 
The (hacky) way that the visualizations are produced is the following. The code, vis_corex.py, produces a text file called "graphs/graph.dot". This just encodes the edges between nodes in dot format. Then, the code calls a command line utility called sfdp that is part of graphviz, 

```
sfdp graph.dot -Tpdf -Earrowhead=none -Nfontsize=12  -GK=2 -Gmaxiter=1000 -Goverlap=False -Gpack=True -Gpackmode=clust -Gsep=0.01 -Gsplines=False -o graph_sfdp.pdf
```

These dot files can also be opened with OmniGraffle if you would like to be able to manipulate them by hand. 
If you want, you can try to recompile graphs yourself with different options to make them look nicer. Or you can edit the dot files to get effects like colored nodes, etc.

For Ubuntu users:

Credits: https://gitlab.com/graphviz/graphviz/issues/1237

1. Remove any existing installation with `conda uninstall graphviz`. (If you did not install with Conda, you might need to do `sudo apt purge graphviz` and/or `pip uninstall graphviz`).
    
2. run `sudo apt install libgts-dev`

3. run `sudo pkg-config --libs gts`
    
4. run `sudo pkg-config --cflags gts`

5. Download `graphviz-2.40.1.tar.gz` from [here](https://graphviz.gitlab.io/pub/graphviz/stable/SOURCES/graphviz.tar.gz)

6. Navigate to directory containing download, and extract with `tar -xvf graphviz-2.40.1.tar.gz` (or newer whatever the download is named.)

7. `cd` into extracted folder (ie `cd graphviz-2.40.1`) and run `sudo ./configure --with-gts`

8. Run `sudo make` in the folder

9. Run `sudo make install` in the folder

10. Reinstall library using `pip install graphviz`
    
 





Changelog.md
## March 25 - 2019

Changes in vis_corex.py

1. codebase Python 2 to Python 3 migration `

2. DeprecationWarning: 'U' mode is deprecated
  
    
    * Before: `with open(filename, 'rU') as csvfile:`  
    * After: `with open(filename, 'r') as csvfile:`


Changes in readme.md

3. added `graphviz` installation instructions for Ubuntu users

4. added `requirements.txt` which lets users to install all the dependencies with a single command `conda install --yes --file requirements.txt`


tests/modules.py
import numpy as np
import sklearn as sk
import sklearn.decomposition as skd
import sklearn.manifold as skm
import sklearn.cluster as skc
import sklearn.neural_network as skn
import sys
sys.path.append('/Users/gregv/deepcloud/LinearSieve')
import linearsieve as sieve
sys.path.append('/Users/gregv/deepcloud/LinearCorex')
import linear_corex as lc
from scipy.stats import kendalltau, pearsonr, spearmanr

methods = [ 
   #(name,type,f)
    ("LC", lambda m, q: lc.Corex(n_hidden=m, verbose=True, max_iter=5000).fit_transform(q)),
   ("Sieve", lambda m, q: sieve.Sieve(n_hidden=m, max_iter=100).fit_transform(q)),
   #("Spectral","c",cluster.SpectralClustering(n_clusters=2, eigen_solver='arpack',affinity="nearest_neighbors")),
   #("LSA","d",TruncatedSVD(n_components=2)),
   # ("RBM", lambda m, q: skn.BernoulliRBM(n_components=m, learning_rate=0.01, n_iter=50).fit_transform(binarize(q))),
   #("NMF", lambda m, q: skd.ProjectedGradientNMF(n_components=m, init='random', random_state=0).fit_transform(-np.min(q) + q)),
   #("Factor Analysis", lambda m, q: skd.FactorAnalysis(n_components=m).fit_transform(q)),
   #("PCA", lambda m, q: skd.PCA(n_components=m).fit_transform(q)),
   #("ICA", lambda m, q: skd.FastICA(n_components=m).fit_transform(q)),
   #("KMeans", lambda m, q: skc.MiniBatchKMeans(n_clusters=m+1).fit_transform(q)[:,:m]),
   #("LLE", lambda m, q: skm.LocallyLinearEmbedding(n_components=m).fit_transform(q)),
   #("Isomap", lambda m, q: skm.Isomap(n_components=m).fit_transform(q)),
   # ("NMF2",ProjectedGradientNMF(n_components=1, init='random', random_state=0)),
   # ("LSA2",TruncatedSVD(n_components=1)),
   # ("PCA2",PCA(n_components=1)),
   # ("ICA3",FastICA(n_components=1)),
   # ("ICA4",FastICA(n_components=1)),
   # ("LLE2",LocallyLinearEmbedding(n_neighbors=3,n_components=1)),
   # ("Isomap2",Isomap(n_neighbors=40,n_components=1)),
   # #("MDSe", MDS(n_components=1,metric=False)),
   # ("Knn",KNeighborsClassifier(3)),
   # ("Linear SVM",SVC(kernel="linear", C=0.025)),
   # ("Decision Tree",DecisionTreeClassifier(max_depth=5)),
   # ("Random Forest",RandomForestClassifier(max_depth=5, n_estimators=10, max_features=1)),
   # ("Adaboost",AdaBoostClassifier()),
   # ("NB",BernoulliNB()),
   # ("Lin. Disc.","s",LDA()),
   # ("Log. Reg.","s",linear_model.LogisticRegression())
   ]

def binarize(q):
  return np.vstack([(x - np.min(x)) / np.ptp(x) for x in q.T]).T

def shannon_k(noise):
  return 1. / (0.5 * np.log(1 + 1./noise))

def shannon_noise(c):
  # Give the noise level compatible with a certain capacity (assuming standard signal)
  return 1. / (np.exp(2. * c) - 1)

def random_frac(k):
  # q = np.random.random(k)
  # return q / np.sum(q)
  return np.random.dirichlet(np.ones(k))

def random_unit(k):
  z = np.random.randn(k)
  return z / np.sqrt(np.dot(z, z))

def gen_data(n_sources=1, k=10, n_samples=10000, noise=0.): # , extra=0):
  """Generate data. The model is that there are n_sources, Z_j, normal with unit variance.
     There are k observed vars per source, X_i = Z_j + E_i. E_i is iid normal noise with variance 'noise' (AWGN). 
     There is a Shannon threshold relating k and noise, defined by shannon_ functions
  """
  sources = np.random.randn(n_samples, n_sources)
  observed = np.vstack([source + np.sqrt(noise) * np.random.randn(n_samples) for source in sources.T for i in range(k)]).T
  # observed = np.hstack([observed, np.random.randn(n_samples, extra)])
  return observed, sources


def gen_data_cap(n_sources=1, k=10, n_samples=1000, capacity=4., scale=False): 
  """Generate data. The model is that there are n_sources, Z_j, normal with unit variance.
     There are k observed vars per source, X_i = Z_j + E_i. E_i is iid normal noise with variance 'noise' (AWGN). 
     Noise is chosen randomly so that the capacity for each source is fixed. 
     There is a Shannon threshold relating k and noise, defined by shannon_ functions
  """
  sources = np.random.randn(n_samples, n_sources)
  capacities = [capacity * random_frac(k) for source in sources.T]
  noises = [shannon_noise(c) for c in capacities]
  observed = np.vstack([source + np.sqrt(noises[j][i]) * np.random.randn(n_samples) for j, source in enumerate(sources.T) for i in range(k)]).T
  if scale:
    scales = 2**(20 * (np.random.random(observed.shape[1]) - 0.5))
    observed = scales * observed  # N by n
  return observed, sources

def gen_data_s(n_sources=1, k=10, n_samples=10000, rho=1): 
  """Generate data. The model is that there are n_sources, Z_j, normal with unit variance.
     There are k observed vars per source, X_i = s_i Z_j + E_i. E_i is iid normal noise with unit variance 'noise' (AWGN). 
  """
  sources = np.random.randn(n_samples, n_sources)
  s = [rho * random_unit(k)  for source in sources.T]
  observed = np.vstack([s[j][i] * source + np.random.randn(n_samples) for j, source in enumerate(sources.T) for i in range(k)]).T
  return observed, sources

def score(true, predicted, method=pearsonr):
    """Compare n true signals to some number of predicted signals.
    For each true signal take the best match in terms of 'method' correlation."""
    rs = []
    for t in true.T:
        rs.append(max(np.abs(method(t, p)[0]) for p in predicted.T))
    return np.mean(rs)

def multi_score(m, C, k, f, n_samples, scale=False):
  results = []
  for i in range(10):
    x, z = gen_data_cap(n_sources=m, k=k, n_samples=n_samples, capacity=C, scale=scale)
    results.append(score(z, f(m, x)))
  return np.mean(results), np.std(results)

def test_solvers(m, C, k, n_samples=1000, scale=False):
  """Use a variety of solvers to recover m sources from data, x."""
  return [(name,) + multi_score(m, C, k, f, n_samples, scale=scale) for name, f in methods]


tests/test_corex.py
from __future__ import print_function
import numpy as np
import sys
sys.path.append('..')
import linear_corex as lc
from scipy.stats import kendalltau


kwargs = {'verbose': True, 'seed': 1}
np.random.seed(kwargs['seed'])
test_array_d = np.repeat([[0, 0, 0], [1, 1, 1]], 3, axis=0)
test_array_c = np.repeat(np.random.random((100, 1)), 3, axis=1)


# CONTINUOUS TESTS
def test_discrete():
    out = lc.Corex(n_hidden=1, **kwargs)
    y = out.fit_transform(test_array_d)
    assert np.allclose(kendalltau(y[:, 0], test_array_d[:, 0]), 1, atol=1e-3)
    out = lc.Corex(n_hidden=1, **kwargs)
    y = out.fit_transform(test_array_c)
    assert np.allclose(kendalltau(y[:, 0], test_array_c[:, 0]), 1, atol=1e-3)


def test_continuous():
    a = np.random.randn(10000, 3)
    a = np.repeat(a, 3, axis=1)
    a += 0.05 * np.random.normal(size=a.shape)
    a -= a.mean(axis=0)
    out = lc.Corex(n_hidden=3, max_iter=10, verbose=True, seed=0).fit(a)
    print(out.ws)
    print(out.tc)


def test_set(groups=(3,2), noise=0.05, samples=1000):
    a = []
    for i, k in enumerate(groups):
        a.append(np.repeat(np.random.randn(samples, 1), k, axis=1))
    a = np.hstack(a)
    a += noise * np.random.normal(size=a.shape)
    a -= a.mean(axis=0)
    return a


tests/test_faces.py
print(__doc__)

# Authors: Vlad Niculae, Alexandre Gramfort
# License: BSD 3 clause

import logging
from time import time
import os
import sys

import numpy as np
from numpy.random import RandomState
import matplotlib.pyplot as plt

from sklearn.datasets import fetch_olivetti_faces
from sklearn.cluster import MiniBatchKMeans
from sklearn import decomposition

sys.path.append('..')
import linear_corex as lc

# Display progress logs on stdout
logging.basicConfig(level=logging.INFO,
                    format='%(asctime)s %(levelname)s %(message)s')

if not os.path.exists('faces'):
    os.makedirs('faces')

n_row, n_col = 4, 5
n_components = n_row * n_col
image_shape = (64, 64)
rng = RandomState(0)

###############################################################################
# Load faces data
dataset = fetch_olivetti_faces(shuffle=True, random_state=rng)
faces = dataset.data

n_samples, n_features = faces.shape

# global centering
faces_centered = faces - faces.mean(axis=0)

# local centering
faces_centered -= faces_centered.mean(axis=1).reshape(n_samples, -1)

print("Dataset consists of %d faces" % n_samples)


###############################################################################
def plot_gallery(title, images, n_col=n_col, n_row=n_row):
    plt.figure(figsize=(2. * n_col, 2.26 * n_row))
    plt.suptitle(title, size=16)
    for i, comp in enumerate(images):
        plt.subplot(n_row, n_col, i + 1)
        vmax = max(comp.max(), -comp.min())
        plt.imshow(comp.reshape(image_shape), cmap=plt.cm.gray,
                   interpolation='nearest',
                   vmin=-vmax, vmax=vmax)
        plt.xticks(())
        plt.yticks(())
    plt.subplots_adjust(0.01, 0.05, 0.99, 0.93, 0.04, 0.)


###############################################################################
# List of the different estimators, whether to center and transpose the
# problem, and whether the transformer uses the clustering API.
estimators = [
    ('Linear Corex',
        lc.Corex(n_hidden=n_components, verbose=True, max_iter=10000), False),
    # ('Eigenfaces - RandomizedPCA',
    #     decomposition.RandomizedPCA(n_components=n_components, whiten=True), True),
    # ('Non-negative components - NMF',
    #     decomposition.NMF(n_components=n_components, init='nndsvda', beta=5.0,
    #                    tol=5e-3, sparseness='components'), False),
    # ('Independent components - FastICA',
    #     decomposition.FastICA(n_components=n_components, whiten=True), True),
    # ('Sparse comp. - MiniBatchSparsePCA',
    #     decomposition.MiniBatchSparsePCA(n_components=n_components, alpha=0.8,
    #                                   n_iter=100, batch_size=3, random_state=rng), True),
    # ('MiniBatchDictionaryLearning',
    #     decomposition.MiniBatchDictionaryLearning(n_components=15, alpha=0.1,
    #                                               n_iter=50, batch_size=3, random_state=rng), True),
    # ('Cluster centers - MiniBatchKMeans',
    #     MiniBatchKMeans(n_clusters=n_components, tol=1e-3, batch_size=20, max_iter=50, random_state=rng), True),
    # ('Factor Analysis components - FA',
    #    decomposition.FactorAnalysis(n_components=n_components, max_iter=2), True),
]


###############################################################################
# Plot a sample of the input data

plot_gallery("First centered Olivetti faces", faces_centered[:n_components])

###############################################################################
# Do the estimation and plot it

for name, estimator, center in estimators:
    print("Extracting the top %d %s..." % (n_components, name))
    t0 = time()
    data = faces
    if center:
        data = faces_centered
    estimator.fit(data)
    train_time = (time() - t0)
    print("done in %0.3fs" % train_time)
    if hasattr(estimator, 'cluster_centers_'):
        components_ = estimator.cluster_centers_
    elif hasattr(estimator, 'components_'):
        components_ = estimator.components_
    else:
        components_ = np.array([estimator.ws[i][:n_features] for i in range(n_components)])
    if hasattr(estimator, 'noise_variance_'):
        plot_gallery("Pixelwise variance",
                     estimator.noise_variance_.reshape(1, -1), n_col=1,
                     n_row=1)
    plot_gallery('%s - Train time %.1fs' % (name, train_time),
                 components_[:n_components])

    plt.savefig('faces/%s.pdf' % name)
    plt.clf()



sys.exit()


n_components = 50
out = lc.Corex(n_hidden=n_components, max_iter=10000)
out.fit(data)
components_ = np.array([out.ws[i][:n_features] for i in range(n_components)])
plot_gallery('%s' % 'Linear Sieve Components', components_[:n_components],  n_col=10, n_row=5)
plt.savefig('faces/big_components.pdf')
plt.clf()

plt.close('all')

covariance/stock.py
from covariance import *
import numpy as np
import cPickle

np.random.seed(1)
prefix = 'figs/'

# Load covariance matrices
dates, dates_string, stocks, x = cPickle.load(open('dates,dates_string,stocks,data.dat'))
x = x - np.mean(x, axis=0, keepdims=True)
x /= np.std(x, axis=0, keepdims=True)
# x = x[:, :100]  # Subset for dev.
print("Data shape is: {}".format(x.shape))
cov = np.cov(x.T)
plot_cov([('Data', x), ('Empirical covariance', cov)], '{}_stock_raw_data'.format(prefix))

m = 60  # Number of factors for FA, CorEx
test_weeks = 26
exp_prefix = '{}stock_{}factor_{}testweeks'.format(prefix, m, test_weeks)
ns = [52 * k for k in range(1, 17)]
p = x.shape[1]
x_train, x_test = x[:-test_weeks], x[-test_weeks:]

results = [(name, []) for name, _ in methods_no_truth]
cov_grid = [[] for _ in range(len(ns))]
for imethod, (name, method) in enumerate(methods_no_truth):
    print('Method: {}'.format(name))
    for ni, n in enumerate(ns):
        print('\tn={}'.format(n))

        mu = np.mean(x_train[-n:], axis=0, keepdims=True)
        stds = np.std(x_train[-n:], axis=0, keepdims=True)
        x_test_scaled = (x_train[-n:] - mu) / stds
        cov = method(m, x_test_scaled) * stds * stds.T
        cov_grid[ni].append(cov)
        results[imethod][1].append(score(x_test, cov))

print 'plot scores'
print results
plot_scores(ns, results, p, name=exp_prefix)

print 'animate matrices'
skip_index = [name for name, _ in methods_no_truth].index("Independent")
names = [name for name, _ in methods_no_truth if name is not 'Independent']
for i in range(len(cov_grid)):
    cov_grid[i] = [cov_grid[i][j] for j in range(len(cov_grid[i])) if j != skip_index]
plot_cov_grid(cov_grid, names, ns, exp_prefix)

covariance/experiments.py
from covariance import *
import numpy as np


np.random.seed(1)
prefix = 'figs/'

# Example covariance matrices
x = np.random.randn(25, 50)
cov = np.cov(x.T)
plot_cov([('True covariance', np.eye(50)), ('Data', x), ('Empirical covariance', cov)], '{}example'.format(prefix))


# EXPERIMENT 1:
p = 256
m = 64  # Number of sources to use for generating data
eps = 0.
capacity = 8
exp_prefix = '{}p={}_m={}_eps={}_c={}'.format(prefix, p, m, eps, capacity)

k_min, k_max = 3, int(np.log2(p)) + 2  # Use between 2**kmin and 2**k_max samples for train or test
ns = [2**k for k in range(k_min, k_max + 1)]
n_max = 2 * 2**k_max  # First half for training, second half for testing
x, true_cov = gen_data_cap(p=p, n_samples=n_max, n_sources=m, correlate_sources=eps, capacity=capacity)
plot_cov([('Ground Truth Covariance', true_cov)], exp_prefix)
x_train, x_test = x[:n_max / 2], x[n_max / 2:]

results = [(name, []) for name, _ in methods]
cov_grid = [[] for _ in range(len(ns))]
for imethod, (name, method) in enumerate(methods):
    print('Method: {}'.format(name))
    for ni, n in enumerate(ns):
        print('\tn={}'.format(n))
        mu = np.mean(x_train[:n], axis=0, keepdims=True)
        stds = np.std(x_train[:n], axis=0, keepdims=True, ddof=0)
        x_test_scaled = (x_train[:n] - mu) / stds
        if name == 'Ground Truth':
            cov = true_cov
        else:
            cov = method(m, x_test_scaled) * stds * stds.T
        cov_grid[ni].append(cov)
        results[imethod][1].append(score(x_test, cov))

print 'plot scores'
print results
plot_scores(ns, results, np.log2(p), name=exp_prefix)

print 'animate matrices'
skip_index = [name for name, _ in methods].index("Independent")
names = [name for name, _ in methods if name is not 'Independent']
for i in range(len(cov_grid)):
    cov_grid[i] = [cov_grid[i][j] for j in range(len(cov_grid[i])) if j != skip_index]
plot_cov_grid(cov_grid, names, ns, exp_prefix)

covariance/covariance.py
from sklearn.covariance import LedoitWolf, OAS, ShrunkCovariance, GraphLassoCV
from sklearn.decomposition import FactorAnalysis
import numpy as np
import matplotlib
# matplotlib.use('MacOSX')
import matplotlib.pyplot as plt
from matplotlib import animation
import sys
from scipy.stats import rankdata
from sklearn.utils.extmath import fast_logdet, pinvh
sys.path.append('..')
import linear_corex as lc
import os
if not os.path.isdir('figs'):
    os.makedirs('figs')

# COVARIANCE MATRIX ESTIMATORS
methods = [
    # (name, f(m, x))  First parameter of f is for parameters (if needed), second is data
    ("Ground Truth", lambda m, x: None),  # Handled with special logic
    ("Independent", lambda m, x: np.eye(x.shape[1])),
    ("Empirical", lambda m, x: np.cov(x.T)),
    ("Ledoit-Wolf", lambda m, x: LedoitWolf(store_precision=False, assume_centered=True, block_size=2000).fit(x).covariance_),
    ("Factor An.", lambda m, x: FactorAnalysis(n_components=m).fit(x).get_covariance()),
    ("LC0", lambda m, x: lc.Corex(n_hidden=m, max_iter=10000, verbose=1, eliminate_synergy=False, gpu=False).fit(x).estimate_covariance()),
    ("GLASSO", lambda m, x: GraphLassoCV().fit(x).covariance_),
    ("LinCorEx", lambda m, x: lc.Corex(n_hidden=m, max_iter=10000, verbose=1, gpu=False).fit(x).estimate_covariance()),
    # ("LinCorExH", lambda m, x: lc.Corex(n_hidden=m, max_iter=10000, verbose=1, gpu=False).fit(x).estimate_covariance_h(x))
]

methods_no_truth = [
    # (name, f(m, x))  First parameter of f is for parameters (if needed), second is data
    ("Independent", lambda m, x: np.eye(x.shape[1])),
    ("Empirical", lambda m, x: np.cov(x.T)),
    ("Ledoit-Wolf", lambda m, x: LedoitWolf(store_precision=False, assume_centered=True, block_size=2000).fit(x).covariance_),
    ("Factor An.", lambda m, x: FactorAnalysis(n_components=m).fit(x).get_covariance()),
    ("LC0", lambda m, x: lc.Corex(n_hidden=m, max_iter=10000, verbose=1, eliminate_synergy=False, gpu=False).fit(x).estimate_covariance()),
    ("GLASSO", lambda m, x: GraphLassoCV().fit(x).covariance_),
    ("LinCorEx", lambda m, x: lc.Corex(n_hidden=m, max_iter=10000, verbose=1, gpu=False).fit(x).estimate_covariance()),
    # ("LinCorExH", lambda m, x: lc.Corex(n_hidden=m, max_iter=10000, verbose=1, gpu=False).fit(x).estimate_covariance_h(x))
]

# DATA GENERATING METHODS
def shannon_noise(c):
    """Give the noise level compatible with a certain capacity (assuming standard signal)."""
    return 1. / (np.exp(2. * c) - 1)


def random_frac(k):
    """Return a random fraction, drawn from a uniform Dirichlet (arranged from highest to lowest)."""
    return np.ones(k) / k
    #return np.sort(np.random.dirichlet(np.ones(k)))[::-1]


def gen_data_cap(p=500, n_samples=500, capacity=4., n_sources=16, correlate_sources=False):
    """Generate data. The model is that there are n_sources, Z_j, normal with unit variance.
     There are k observed vars per source, X_i = Z_j + E_i. E_i is iid normal noise with variance 'noise' (AWGN).
     Noise is chosen randomly so that the capacity for each source is fixed.
     There is a Shannon threshold relating k and noise, defined by shannon_ functions
    """
    print("Correlate sources:{}".format(correlate_sources))
    if correlate_sources:
        eps = correlate_sources
        sources = (eps * np.random.randn(n_samples, 1) + np.random.randn(n_samples, n_sources)) / np.sqrt(1 + eps**2)
    else:
        sources = np.random.randn(n_samples, n_sources)
    k = p / n_sources
    assert p % n_sources == 0, 'For simplicity, we force k variables per source.'
    capacities = [capacity * random_frac(k) for _ in sources.T]
    noises = [shannon_noise(c) for c in capacities]
    observed = np.vstack([(source + np.sqrt(noises[j][i]) * np.random.randn(n_samples)) / np.sqrt(1 + noises[j][i])
                          for j, source in enumerate(sources.T) for i in range(k)]).T
    cov = np.eye(p)
    # for j in range(n_sources):
    #     for i in range(k):
    #         for ip in range(k):
    #             if i != ip:
    #                 cov[i + k * j, ip + k * j] = 1 / np.sqrt(1 + noises[j][i]) / np.sqrt(1 + noises[j][ip])
    print 'maximum off diagonal is', 1 / np.sqrt(1 + noises[0][0]) / np.sqrt(1 + noises[0][1])
    for i in range(p):
        for ip in range(p):
            if i == ip:
                cov[i, ip] = 1
            elif i / k == ip / k:
                cov[i, ip] = 1 / np.sqrt(1 + noises[i/k][i%k]) / np.sqrt(1 + noises[i/k][ip%k])
            elif correlate_sources:
                cov[i, ip] = eps**2 / (1 + eps**2)
    return observed, cov


# SCORING METHOD
def score(test_data, cov):
    """Use negative log-likelihood to evaluate."""
    test_cov = np.cov(test_data.T)
    result = -log_likelihood(test_cov, cov) / len(test_data)  # Convert units to bits / test sample
    return np.clip(result, -1e6, 1e6)


def log_likelihood(emp_cov, model_cov):
    """Computes the sample mean of the log_likelihood under a covariance model
    computes the empirical expected log-likelihood (accounting for the
    normalization terms and scaling), allowing for universal comparison (beyond
    this software package)
    Parameters
    ----------
    emp_cov : 2D ndarray (n_features, n_features)
        Maximum Likelihood Estimator of covariance
    precision : 2D ndarray (n_features, n_features)
        The precision matrix of the covariance model to be tested
    Returns
    -------
    sample mean of the log-likelihood
    """
    p = model_cov.shape[0]
    precision = pinvh(model_cov)
    log_likelihood_ = - np.sum(emp_cov * precision) + fast_logdet(precision)
    log_likelihood_ -= p * np.log(2 * np.pi)
    log_likelihood_ /= 2.
    return log_likelihood_


# PLOTTING METHODS
tableau = ['#1f77b4', '#ff7f0e', '#2ca02c', '#d62728', '#9467bd', '#8c564b', '#e377c2', '#7f7f7f', '#bcbd22', '#17becf']


def plot_scores(xs, scores, p, name='plot', option=None):
    plt.style.use('https://dl.dropboxusercontent.com/u/23723667/gv.mplstyle')
    #plt.style.use('/Users/gregv/Dropbox/Public/gv.mplstyle')
    fig = plt.figure(figsize=(8, 5))
    ax = plt.subplot(111)

    ax.yaxis.set_label_position("right")
    plt.ylabel('Negative Log Likelihood\n (avg per test sample)', fontsize=18, fontweight='bold')
    plt.xlabel('n samples (log scale)', fontsize=18, fontweight='bold')

    # Limits
    ymin = np.min([score for _, score in scores])
    ind_index = [n for n, score in scores].index('Independent')
    ymax = np.max(scores[ind_index][1])  # Independent is Baseline
    buffer = (ymax - ymin) / 5.
    y0, y1 = float(int((ymin - buffer)*100)) / 100., np.around(ymax + buffer, decimals=2)
    dy = 10.**int(np.log10(buffer))
    if dy < buffer / 5:
        dy *= 2
    plt.ylim(y0, y1)
    if np.allclose(np.log2(xs), np.around(np.log2(xs))):
        xs = np.log2(xs)  # Log scale
        plt.xticks(xs, ['$2^{%d}$' % x for x in xs], fontsize=16)
    x0, x1 = xs[0], xs[-1]
    plt.xlim(x0, x1)

    # Make sure your axis ticks are large enough to be easily read.
    plt.yticks(np.arange(y0, y1, dy))


    # Add bars
    # for y in np.arange(y0+1, y1, dy):
    #    plt.plot([x0, x1], [y, y], "--", lw=0.5, color="black", alpha=0.3)
    #for x in xs:
    #    plt.plot([x, x], [y0, y1], "--", lw=0.5, color="black", alpha=0.3)
    plt.plot([p, p], [y0, y1], ":", lw=2., color="red", alpha=0.8)  # n=p, special color
    plt.text(p, y0, ' $n=p$', fontsize=16, fontweight='bold', color='red')

    if option is not None:
        pass

    # Remove the tick marks; they are unnecessary with the tick lines we just plotted.
    plt.tick_params(axis="both", which="both", bottom="off", top="off",
                    labelbottom="on", left="off", right="off", labelleft="off", labelright='on')

    y_pos = rankdata([errs[0] for _, errs in scores], method='ordinal').astype(float) / len(scores) * (y1 - y0) + y0
    y_pos = (np.array([np.clip(errs[0], y0, y1) for _, errs in scores]) + y_pos)/2
    for j, (method, errs) in enumerate(scores):
        print method, errs
        if method in ["Ground Truth", "Independent", "Empirical"]:
            valign = {"Ground Truth": 'top', "Independent": "bottom", "Empirical": "top"}[method]
            plt.plot(xs, errs, '-', lw=3.5, color=tableau[j])
            plt.text(xs[-1], y_pos[j], method, fontsize=18, fontweight='bold',
                     verticalalignment=valign, horizontalalignment='right', color=tableau[j])
        else:
            plt.plot(xs, errs, '-', lw=2.5, color=tableau[j])
            plt.text(xs[0], y_pos[j], method, fontsize=18, fontweight='bold',
                 verticalalignment='center', horizontalalignment='right', color=tableau[j])

        plt.savefig("{}_build{}.png".format(name, j))

    plt.tight_layout()
    plt.savefig("{}.png".format(name))
    plt.close('all')


def plot_cov(covs, name='cov'):
    plt.style.use('https://dl.dropboxusercontent.com/u/23723667/gv.mplstyle')
    # plt.style.use('/Users/gregv/Dropbox/Public/gv_heat.mplstyle')
    colorscheme = plt.cm.RdBu_r

    # Print key
    plt.figure(figsize=(5, 1))
    plt.imshow([[-1, 1]], cmap=colorscheme, interpolation='none', vmin=-1, vmax=1)
    plt.gca().set_visible(False)
    cb = plt.colorbar(orientation='horizontal', ticks=[])
    cb.outline.set_edgecolor('white')
    plt.savefig("figs/key.png")
    plt.clf()

    # Plot matrices
    plt.figure(figsize=(15, 16 * len(covs)), frameon=False)
    for i in range(len(covs)):
        ax = plt.subplot(1, len(covs), 1 + i)
        cov_name, cov = covs[i]
        ax.set_title(cov_name)
        ax.set_axis_off()
        ax.imshow(cov, interpolation='none', cmap=colorscheme, vmin=-1, vmax=1)

    plt.savefig("{}_cov.png".format(name))
    plt.close('all')

def plot_cov_grid(cov_grid, names, ns, name='grid'):
    colorscheme = plt.cm.RdBu_r
    matplotlib.rcParams.update(matplotlib.rcParamsDefault)  # Ever using "tight layout" messes up all future animations

    # Plot matrices
    n_rows = 2
    n_cols = int(np.ceil(float(len(names)) / n_rows))
    n = cov_grid[0][0].shape[0]
    fig, axs = plt.subplots(n_rows, n_cols, figsize=(4.5 * n_cols, 5 * n_rows))
    title = fig.suptitle('', fontsize=18, fontweight='bold')

    artists = []
    for i, ax in enumerate(axs.ravel()):
        if i < len(names):
            ax.set_title(names[i])
        ax.set_axis_off()
        artist = ax.imshow(np.zeros((n, n)), interpolation='nearest', cmap=colorscheme, vmin=-1, vmax=1)
        artists.append(artist)

    def init():
        for i, artist in enumerate(artists):
            artist.set_array(np.zeros((n, n)))
        title.set_text('')
        return artists + [title]

    def animate(j):
        for i, artist in enumerate(artists):
            if i < len(names):
                artist.set_array(cov_grid[-j - 1][i])
        title.set_text('n = {}'.format(ns[-j-1]))
        return artists + [title]

    anim = animation.FuncAnimation(fig, animate, init_func=init, frames=len(cov_grid), blit=True, repeat=False)
    anim.save("{}_grid.mp4".format(name), fps=0.5, extra_args=['-pix_fmt', 'yuv420p', '-vcodec', 'libx264'])
